moz_pioneer/io_modules/decoders/moz_pioneer/jose.lua

```diff
@@ -0,0 +1,342 @@
+        local mode = lfs.attributes(fqdn, "mode")
+        if mode == "directory" and not dn:match("^%.") then
+            for fn in lfs.dir(fqdn) do
+                local name, version = fn:match("(.+)%.(%d+).schema%.json$")
+                if name then
+                    local fh = assert(io.input(string.format("%s/%s", fqdn, fn)))
+                    local schema = fh:read("*a")
+                    local s = schemas[name]
+                    if not s then
+                        s = {}
+                        schemas[name] = s
+                    end
+                    local ok, rjs = pcall(rjson.parse_schema, schema)
+                    if not ok then error(string.format("%s: %s", fn, rjs)) end
+                    s[tonumber(version)] = rjs
+                end
+            end
+        end
+    end
+end
+load_schemas()
+
+
+--[[
+Read the raw message, annotate it with our error information, and attempt to inject it.
+--]]
+local function inject_error(hsr, err_type, err_msg, extra_fields)
+    local len
+    local raw = hsr:read_message("raw")
+    local err = decode_message(raw)
+    err.Logger = nil
+    err.Type = "pioneer.error"
+    if not err.Fields then
+        err.Fields = {}
+    else
+        len = #err.Fields
+        for i = len, 1, -1  do
+            local name = err.Fields[i].name
+            if name == "X-Forwarded-For" or name == "remote_addr" then
+                table.remove(err.Fields, i)
+            end
+        end
+    end
+    len = #err.Fields
+    len = len + 1
+    err.Fields[len] = { name="DecodeErrorType", value=err_type }
+    len = len + 1
+    err.Fields[len] = { name="DecodeError",     value=err_msg }
+
+    if extra_fields then
+        -- Add these optional fields to the raw message.
+        for k,v in pairs(extra_fields) do
+            len = len + 1
+            err.Fields[len] = { name=k, value=v }
+        end
+    end
+    pcall(inject_message, err)
+end
+
+
+local submissionField = {value = nil, representation = "json"}
+local doc = rjson.parse("{}") -- reuse this object to avoid creating a lot of GC
+local function process_json(hsr, msg, schema)
+    local ok, jwe = pcall(jose.jwe_import, hsr:read_message("Fields[content]"))
+    if not ok then
+        inject_error(hsr, "jwe_import", jwe, msg.Fields)
+        return false
+    end
+
+    local ok, json = pcall(jwe.decrypt, jwe, jwk)
+    if not ok then
+        inject_error(hsr, "jwe_decrypt", json, msg.Fields)
+        return false
+    end
+
+    local ok, err = pcall(doc.parse, doc, json)
+    if not ok then
+        inject_error(hsr, "json", string.format("invalid submission: %s", err), msg.Fields)
+        return false
+    end
+
+    ok, err = doc:validate(schema)
+    if not ok then
+        inject_error(hsr, "json", string.format("%s schema version %d validation error: %s", msg.Fields.docType, msg.Fields.sourceVersion, err), msg.Fields)
+        return false
+    end
+
+    submissionField.value = doc
```

Comment:
As an optimization I would recommend removing the JSON encoding entirely and expanding the structure into the native protobuf message.